### DISCUSSION TITLE
some feature requests

### DISCUSSION THREAD

**Discussion by vector65536** [thevindu-w/clip_share_client#10] [Ideas]
just discovered this app. nice work!! feel no pressure to implement some features:

on the client side: 
- shortcut to launch destination folder
- set custom destination (different folder / ext sd card), show basic storage status (free / total)
- small thumbnails for received files (if applicable)
- last seen / online status (green/red color) for target PC
- show target by PC / machine name, in addition to IP
- run minimized (if app is active: show persistent notification with options: `get`, `send`, `exit`)

on the server side:
- autostart (or just copy shortcut to the startup folder: win+R: `shell:startup`)
- show active clients by name/IPs

common:
- auto close if idle for xx minutes (app is lightweight but good to have this option)
- NAT traversal
- add a parameter `trusted` for target PC/android, to auto send clipboard to specific (trusted) machines.

**Comment by thevindu-w** [comment]
Thank you for your suggestions. I'll try to get as many of the features included in future releases.

> shortcut to launch destination folder

This seems a useful feature. I guess you meant to give an option to open the file manager to view the files received. This shouldn't take much time, so, I think I can get this feature implemented in next release.

> custom destination storage status

I'm not clear what this means. Is that an indicator to show the free space on the internal storage? Or is it a way to select a different location to store the received files?

> small thumbnails for received files

This might take some time. But I'll try to make it as soon as possible.

> target icon and color status (green/red)

I'm not clear what this means. What did you mean by a target icon?

> show target by PC / machine name in addition to IP

This is also useful. I already have it in my TODO list, so, this will be implemented soon.

<br>

> autostart

This requires some sort of installation. The current way ClipShare server works has no installation phase. But it's not hard to have using some `cmd` or `sh` scripts. So I'll create some scripts to do the auto-start.

> show clients

I guess you meant showing the currently connected clients. Actually the clients don't keep the connection for a long period (unless it gets a background sync feature). It creates a new connection to perform a task (ex: Get copied text), and once it's done, it will close the connection. Also, there's no GUI for the server (at least per now). The only graphical indication that the server is running, is available only on Windows (the tray icon). I'll try to show the currently connected clients (or the clients connected in the last few minutes) there if possible.

<br>

> auto close if idle for xx minutes

This is possible. I'll include this in my pending feature plans. But even if it does keep open, it won't consume much resources (ex: battery, memory)

> NAT traversal

In fact, this was possible in earlier versions using a reverse tunnel (it also needed an additional supporting program running on the server which I didn't release). I *temporarily* removed it as it's implementation had several vulnerabilities and I thought no one would be interested in such feature (assuming both the client and server are in the same WLAN so there's no NAT between them). But since you are interested to have that feature, I'll get it back in my planned features.

> `trusted` target parameter to auto send clipboard

auto-send clipboard requires the Android client to access the clipboard in the background. But I had some issues with that as newer Android versions restricts background clipboard access (I have mentioned this problem in another issue comment in the server repo, and one suggested some ways to overcome that). If I can get background clipboard access working, it will automatically include this feature.

**Reply by vector65536** [reply]
thanks for the response. to elaborate my thread, here's what I meant (with highlighted some others):

on the client side: 
- shortcut to launch destination folder (or at least, list existing items, with greyed out deleted items since last receival)
- set custom destination (different folder / ext sd card), also show basic storage status (free / total)
- small thumbnails for received files (if applicable), greyed out or x mark for deleted since receival.
- last seen / online status (green/red color) for target PC
- show target by PC name / machine name, in addition to IP
- **run minimized (if app is active: show persistent notification with options: `get`, `send`, `exit`)**

on the server side:
- autostart **(or just copy shortcut to the startup folder: win+R: `shell:startup`)**
- show active clients by name/IPs
- **import / export configurations. plus: save the config file together with exe (or in a predefined location).**

common:
- auto close if idle for xx minutes (app is lightweight but would be nice to have this option)
- **retain clipboard (text) history for n_days or upto n_items.**
- NAT traversal
- add a parameter `trusted` for target PC/android, to auto send clipboard to specific (trusted) machines.

**Reply by thevindu-w** [reply]
> set custom destination (different folder / ext sd card), also show basic storage status

This should also be possible but it may require additional storage permissions. I'll try to include that.

> last seen / online status (green/red color) for target PC

Now I got it. This means a connection history with PC names and last connected date/time. I'll add it in a future release.

> run minimized (if app is active: show persistent notification with options: get, send, exit)

I'll check this also. I think I can include this in next few releases unless the background clipboard access restriction applies for this.

> import / export configurations. plus: save the config file together with exe

If you are not using the default configurations, the config should already be in a file named `clipshare.conf`. If you refer to the default config, you can download it from the [Releases](https://github.com/thevindu-w/clip_share_server/releases) section of the server. If I haven't understood your requirement properly, please correct me.

> retain clipboard (text) history for n_days or upto n_items.

On Android, doesn't this automatically happen in the system's clipboard unless you clear them? Also Windows has the clipboard history feature built in. But I haven't checked the clipboard history on Windows with ClipShare. I'll check that one.

**Reply by vector65536** [reply]
that sums it up. thank you

**Reply by qupear** [reply]
Hi. Sorry to bother, but could you please confirm, that app cannot share between two PC? Only between Android and PC? Could you recommend something similar to share clipboard between PCs? I know LocalSend but you need to open window and paste there. Which is inconvenient for text sharing.

**Reply by thevindu-w** [reply]
> could you please confirm, that app cannot share between two PC?

It's possible. Please refer to [github.com/thevindu-w/clip_share_client_desktop](https://github.com/thevindu-w/clip_share_client_desktop) which is a desktop client for ClipShare. Also I'm planning to create another better and lighter desktop client which is currently under development.

**Comment by mefleacuzimtiny** [comment]
Hi. I've been using the app lately and it is genuinely extremely helpful. So much so that I recommended it to all my friends. One thing I would like to see if it's not too difficult to implement would be a notification menu icon... kinda like the wifi icon and the mobile data icon when you swipe down (I don't know what it's called). I was thinking I could then replace the nearby share button that I rarely use with clipshare.

Furthermore, it'd also be nice if said button opened up a persistent notification to send and get stuff without having to open the actual app separately like a previous comment suggested. Alternatively, this (or maybe a separate) button could also be toggled to constantly listen for clipboard events and then send items to desktop automatically.

**Reply by thevindu-w** [reply]
Thanks for your feedback. I checked the features you suggested. 

> kinda like the wifi icon and the mobile data icon when you swipe down (I don't know what it's called). I was thinking I could then replace the nearby share button that I rarely use with clipshare.

I guess you meant Quick Settings buttons. Typically, they are not used to open apps. However it should be possible to open ClipShare from a quick settings button/tile. I will try to include this feature if I get it implemented without any issue.

> ... send and get stuff without having to open the actual app ...

This requires accessing the clipboard from the background. There are strict restrictions on clipboard access in Android. Most newer Android versions doesn't let apps access clipboard while in the background. I found a workaround for this which requires the app to be opened temporarily until it creates the notification, and then the notification action buttons can access the clipboard even after closing the app. However I have to check if this method works on the latest Android versions, and if there are any other issues with this method (whether this notification can access the clipboard long time after closing the app). If this method works as expected, I'll add this as a feature.
